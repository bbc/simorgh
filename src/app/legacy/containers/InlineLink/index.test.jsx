import React from 'react';
import { StaticRouter } from 'react-router-dom';
import {
  render,
  fireEvent,
} from '../../../components/react-testing-library-with-providers';
import InlineLinkContainer from './index';

const fragmentBlock = (text, attributes = []) => ({
  id: '91238901',
  type: 'fragment',
  model: {
    text,
    attributes,
  },
});

// eslint-disable-next-line react/prop-types
const InlineLinkContext = ({ locator, isExternal, blocks, onClick }) => (
  <StaticRouter>
    <InlineLinkContainer
      locator={locator}
      blocks={blocks}
      isExternal={isExternal}
      onClick={onClick}
    />
  </StaticRouter>
);

describe('InlineLinkContainer', () => {
  describe('link matching routes for SPA', () => {
    it('should render correctly', () => {
      const { container } = render(
        /*
      for the value it would bring, it is much simpler to wrap a react-router Link in a Router, rather than mock a Router or pass some mocked context.
    */
        <StaticRouter>
          <InlineLinkContainer
            locator="https://www.bbc.com/news/articles/c0g992jmmkko"
            blocks={[fragmentBlock('This is text for an internal link')]}
            isExternal={false}
          />
        </StaticRouter>,
      );
      expect(container).toMatchSnapshot();
    });

    it('should render correctly for TEST environment', () => {
      const { container } = render(
        /*
      for the value it would bring, it is much simpler to wrap a react-router Link in a Router, rather than mock a Router or pass some mocked context.
    */
        <StaticRouter>
          <InlineLinkContainer
            locator="https://www.test.bbc.com/news/articles/c0g992jmmkko"
            blocks={[fragmentBlock('This is text for an internal link')]}
            isExternal={false}
          />
        </StaticRouter>,
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('internal link not matching SPA route', () => {
    it('should render correctly', () => {
      const { container } = render(
        <InlineLinkContainer
          locator="https://www.bbc.com/news"
          blocks={[fragmentBlock('This is bold text for a link', ['bold'])]}
          isExternal={false}
        />,
        { service: 'news' },
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('external link accessibility', () => {
    it('should be explicitly marked "external" for screen reader users', () => {
      const { container } = render(
        <InlineLinkContainer
          locator="https://www.example.com/"
          blocks={[fragmentBlock('This is a link')]}
          isExternal
        />,
        { service: 'news' },
      );
      expect(container).toMatchSnapshot();
    });

    it('should be explicitly marked "external" for screen reader users & localised', () => {
      const { container } = render(
        <InlineLinkContainer
          locator="https://www.example.com/"
          blocks={[fragmentBlock('این لینک هست')]}
          isExternal
        />,
        { service: 'persian' },
      );
      expect(container).toMatchSnapshot();
    });

    describe('onClick', () => {
      const mockOnClick = jest.fn();
      afterEach(() => {
        jest.clearAllMocks();
      });
      describe('onClick event on links', () => {
        it('should send event when onClick is not undefined', () => {
          const { getByText } = render(
            <InlineLinkContext
              locator="https://www.bbc.com/news"
              isExternal={false}
              blocks={[fragmentBlock('This is a link')]}
              onClick={mockOnClick}
            />,
          );

          const linkButton = getByText('This is a link');

          expect(mockOnClick.mock.calls.length).toBe(0);

          fireEvent.click(linkButton, { button: 0 });

          expect(mockOnClick.mock.calls.length).toBe(1);
        });

        it('should not send event when onClick is undefined', () => {
          const { getByText } = render(
            <InlineLinkContext
              locator="https://www.bbc.com/news"
              isExternal={false}
              blocks={[fragmentBlock('This is a link')]}
              onClick={undefined}
            />,
          );

          const linkButton = getByText('This is a link');

          fireEvent.click(linkButton, { button: 0 });

          expect(mockOnClick.mock.calls.length).toBe(0);
        });
      });
    });
  });
});
