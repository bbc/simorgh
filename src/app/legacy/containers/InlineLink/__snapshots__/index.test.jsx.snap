// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InlineLinkContainer external link accessibility should be explicitly marked "external" for screen reader users & localised 1`] = `
.emotion-0 {
  color: #222222;
  border-bottom: 1px solid #B80000;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:visited {
  color: #6E6E73;
  border-bottom: 1px solid #6E6E73;
}

.emotion-0:focus,
.emotion-0:hover {
  border-bottom: 2px solid #B80000;
  color: #B80000;
}

<div>
  <a
    aria-label="این لینک هست، خارجی"
    class="focusIndicatorReducedWidth emotion-0 emotion-1"
    href="https://www.example.com/"
  >
    این لینک هست
  </a>
</div>
`;

exports[`InlineLinkContainer external link accessibility should be explicitly marked "external" for screen reader users 1`] = `
.emotion-0 {
  color: #222222;
  border-bottom: 1px solid #B80000;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:visited {
  color: #6E6E73;
  border-bottom: 1px solid #6E6E73;
}

.emotion-0:focus,
.emotion-0:hover {
  border-bottom: 2px solid #B80000;
  color: #B80000;
}

<div>
  <a
    aria-label="This is a link, external"
    class="focusIndicatorReducedWidth emotion-0 emotion-1"
    href="https://www.example.com/"
  >
    This is a link
  </a>
</div>
`;

exports[`InlineLinkContainer internal link not matching SPA route should render correctly 1`] = `
.emotion-0 {
  color: #222222;
  border-bottom: 1px solid #B80000;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:visited {
  color: #6E6E73;
  border-bottom: 1px solid #6E6E73;
}

.emotion-0:focus,
.emotion-0:hover {
  border-bottom: 2px solid #B80000;
  color: #B80000;
}

<div>
  <a
    class="focusIndicatorReducedWidth emotion-0 emotion-1"
    href="/news"
  >
    <b>
      This is bold text for a link
    </b>
  </a>
</div>
`;

exports[`InlineLinkContainer link matching routes for SPA should render correctly 1`] = `
.emotion-0 {
  color: #222222;
  border-bottom: 1px solid #B80000;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:visited {
  color: #6E6E73;
  border-bottom: 1px solid #6E6E73;
}

.emotion-0:focus,
.emotion-0:hover {
  border-bottom: 2px solid #B80000;
  color: #B80000;
}

<div>
  <a
    class="focusIndicatorReducedWidth emotion-0 emotion-1"
    href="/news/articles/c0g992jmmkko"
  >
    This is text for an internal link
  </a>
</div>
`;

exports[`InlineLinkContainer link matching routes for SPA should render correctly for TEST environment 1`] = `
.emotion-0 {
  color: #222222;
  border-bottom: 1px solid #B80000;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-0:visited {
  color: #6E6E73;
  border-bottom: 1px solid #6E6E73;
}

.emotion-0:focus,
.emotion-0:hover {
  border-bottom: 2px solid #B80000;
  color: #B80000;
}

<div>
  <a
    class="focusIndicatorReducedWidth emotion-0 emotion-1"
    href="/news/articles/c0g992jmmkko"
  >
    This is text for an internal link
  </a>
</div>
`;
